Features/deletedata: Create Privacy Controls UI and Add deleteFlag to specified timestamp

carolinentran opened the pull request:
# *Add deleteFlag to a specified timestamp in Firestore*

## :recycle: Current situation & Problem
While the current version of the app allows reading in health data to be disabled via toggles in ToggleTestView.swift, it does not allow past health data to be reviewed and hidden or deleted. This PR creates a page ManageDataView for users to be able to see all health data types (e.g, step count) with toggle status shown as well as detailed pages per datatype in DeleteDataView that allows users to switch toggle status (enable/disable reading data type) and hide data from a list of timestamps. A function addDeleteFlag was created to add a deleteFlag and set it to true for a timestamp the user chooses to hide. 


## :gear: Release Notes 
Updates
- Created ManageDataView: A list of all healthkit data sample types that link to a DeleteDataView screen specific to the selected category.  
- Created new file DeleteDataView: Using the dictionaries in PrivacyModule we created a details screen showing the disable/enable status and a hardcoded array of timestamps (of collected data points). This array will be replaced by an array populated by the actual timestamps from firestore in the next PR. The user will navigate to this screen to hide/unhide collected data. 
- Inside Privacy Module: Created dictionaries iconsMapping, togglesMap, and identifierUIString to map data sample types to icon image, toggle status, and UI title respectively. 
- Inside PrismaStandard+HealthKit.swift, created func addDeleteFlag(quantityType: String, timestamp: String) async that takes in a quantityType (e.g, stepcount, heartrate) and timestamp, uses getPath, and uses setData to add and set a new flag “deleteFlag” to true without replacing original fields.

Next Steps
- While the timestamps in the listview UI were hard-coded, in the next few days, we will query the 10 most recent timestamps from firestore using order and limit data, store it into an array, and display them (allowing the user to also load in more sets of timestamps in increments of 10).
- Change "delete" to "hide" data for adding deleteFlag
- Allow users to delete data view custom range (they will input start and end date for deletion).


## :books: Documentation
N/A


## :white_check_mark: Testing
On HCI-Prisma repository, run `firebase emulators:start --import=firestore_export` to run the firebase emulator. Erase all contents and settings from the device on the simulator. Log in with “[test@account.com](mailto:test@account.com)” and “testing123”. If a timestamp is deleted on DeleteDataView.swift, a new field “deleteFlag” set to true should be created inside the matching timestamp document in the emulator.

## :pencil: Code of Conduct & Contributing Guidelines 

By submitting creating this pull request, you agree to follow our [Code of Conduct](https://github.com/CS342/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/CS342/.github/blob/main/CONTRIBUTING.md):
- I agree to follow the [Code of Conduct](https://github.com/CS342/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/CS342/.github/blob/main/CONTRIBUTING.md).

## Roles:

- Evelyn: We both worked together on each step but I focused on creating the ManageDataView page to display a list of data type sources and disabled/enabled status using a new struct DataCategoryItem; creating environment accessible dictionaries in Privacy Modules to map data sample types to toggle status, icon, and title; UI for displaying list and animating deletion of timestamps in DeleteDataView; creating addDeleteFlag to create and set new field deleteFlag to true in corresponding timestamp in Firestore; and debugging addDeleteFlag to fix out-of-index and incorrect getPath errors.
- Caroline: We both worked together on each step but I focused on the linking between the screens (Home and navigation bar, ManageDataView to DeleteDataView, communication between functions and dictionaries in Privacy Modules); creating the user action trigger that writes a deleteFlag into Firestore; restructuring our logic and functionality in DeleteDataView to use getPath() and HKSamples; bug fixes for the UI on both ManageDataView and DeleteDataView; and bug fixes to the addDeleteFlag and UI item remove animation so the item is flagged in the backend before the deletion is shown to the user. 
- Dhruv: I focused on implementing certain functionalities throughout the entire UI development process. This included providing the initial architecture for the PrivacyModule class, adding toggle functionality for a users to omit a certain data type from being collected within our new UI post-midterm, and writing functions to set up this toggle creation by altering the PrismaStandard class and the PrivacyModule class to take in a list of HKSampleTypes and create a Binding set of Toggles which map identifiers to the Boolean toggle’s state.

evelyn-hur committed:
Add delete flag

Add delete flag for specified time stamp

evelyn-hur committed:
Add UI for deleting data

Made SwiftUI table to display all recorded health data. Tried to call addDeleteFlag function but ran into async errors.

carolinentran committed:
Linking a privacy tab to privacy controls page

I created a privacy controls tab in the navigation bar which is linked to a list of items for redaction.

Co-Authored-By: Evelyn Hur <122575337+evelyn-hur@users.noreply.github.com>
Co-Authored-By: Bryant Jimenez <84168212+bryant-jimenez@users.noreply.github.com>

carolinentran committed:
get timestamps and set flag (buggy)

wrote a function that gets all the timestamps for the user. (bug: getPath requires a quantityType that we are not able to fully incorporate yet)

wrote a function that sets the delete flag (bug: not able to pass in a specific timestamp just yet)

Co-Authored-By: Evelyn Hur <122575337+evelyn-hur@users.noreply.github.com>
Co-Authored-By: Bryant Jimenez <84168212+bryant-jimenez@users.noreply.github.com>

mjoerke committed:
adding sleep and workout types

mjoerke committed:
Fix linting errors

carolinentran committed:
Created new function for getting user path

The userCustomGetPath was a modification of the getPath() function which gets the path for the user who is currently logged in. This allows for the quantity type and timestamp values to be passed in as strings pulled from the UI.

mjoerke commented on Prisma/Standard/PrismaStandard.swift:
`userCustomGetPath(moduleText: String)` is not needed. You can call `getPath(module: .health(identifier))`, which will give you the same result.

mjoerke replied:
See `PrismaStandard+HealthKit.swift` where we define the upload path using
`path = try await getPath(module: .health(identifier)) + "raw/\(effectiveTimestamp)"`

carolinentran replied:
I just made the modification but I currently have the stepcount HKQuantityType hardcoded. I'll work with @evelyn-hur to create an array of selection options that allow the user to choose different HKQuantityTypes. We may have to map these different HKQuantityTypes to a string value (the string shown on the UI) in order to identify which was selected.

carolinentran committed:
Modification to use HKQuantityType

Rather than using the modified getPath function which takes in a string, I modified the type passed from the deleteDataView to be HKQuantityType instead of a string. The type stepcount is hardcoded in right now, I hope to add logic to pull different HKQuantityType types depending on user selection in the category of data redaction.

evelyn-hur committed:
merged sleep and workouts with deleteData

evelyn-hur committed:
First attempt at toggle mapping identifier string to bool

Created function createTogglesMapping in Privacy Controls module. Previously, sample types (e.g, "step count") were hardcoded to toggle bool. Currently trying to use getSampleIdentifier to map identifier string to toggle bool. We have a bug where we can't access prisma delegate healthKit since it's private.

carolinentran committed:
Identifier Dictionaries for Manage Data Screens

Creating dictionaries that holds information for the ManageData categories (identifier and icon). These dictionaries can also be used to create the specific DeleteDataView screens by passing in the identifier or some information about the specific category selected.

evelyn-hur committed:
Add ManageDataView.swift

Add ManageDataView.swift for user to manage all health data

evelyn-hur committed:
Merge branch 'features/deletedata' of https://github.com/CS342/2024-Prisma into features/deletedata

evelyn-hur committed:
Edited ManageDataView

Display list of datasources with name (e.g, step count), iconName, and toggle status (enabled/disabled) in ManageDataView

carolinentran committed:
Using Dictionary in Private Module for ManageDataView

Using the toggleMap and iconsMapping dictionaries, create an array of items to pass into the ManageDataView.

Current Bug: "Thread 1: Fatal error: No Observable object of type PrivacyModule found. A View.environmentObject(_:) for PrivacyModule may be missing as an ancestor of this view."

mjoerke committed:
Quick fix for runtime error – due to not registering Module in the SpeziAppDelegate configuration

carolinentran committed:
Add additional info to ManageDataView and communicate with DeleteDataView

Added enabled status and icon to ManageDataView

Added navigation from ManageDataView selection to DeleteDataView by passing selected data type identifier

Cleaned up DeleteDataView to loop through all deleted items and set a flag for each deleted item in firestore. Currently still using a static array for timestamps. Next PR will address reading in that data.

dhruvna1k committed:
added toggle based on data type + other small cleanups

evelyn-hur committed:
Fixed DeleteDataView UI

Fixed DeleteDataView spacing, padding, titles, and text.

dhruvna1k committed:
pulled out HKSampleType list and wrote another getIdentifier function

evelyn-hur committed:
Fixed addDeleteFlag in DeleteDataView

Fixed adding delete flag in DeleteDataView by modifying .healthKitDescription to return self by default (temporary fix). Also fixed bug where deleting last timestamp in array returns "index out of range."

evelyn-hur committed:
Sort datatype dictionary into array of tuples

Sort datatype dictionary into array of key-value tuples

evelyn-hur committed:
Fixed lint errors

Fixed lint errors

dhruvna1k committed:
used configure() to populate updated toggleMap

carolinentran committed:
Removed userCustomGetPath

userCustomGetPath() was originally created to accommodate for the missing module type and identifier string due to a previous bug. We were able to take out this custom function and use the general getPath() instead.

dhruvna1k reviewed (approved):
Looks good!

bryant-jimenez commented on Prisma/PrivacyControls/ToggleTestView.swift:
We can likely go ahead and remove this file now that we have a working data view UI @carolinentran @evelyn-hur @dhruvna1k
